examples/cfd/stokes_mgn/pi_inference.py

```diff
@@ -0,0 +1,525 @@
+        results_int = self.graph_full.forward({"x": x_int, "y": y_int})
+        pred_mom_u, pred_mom_v, pred_cont = (
+            results_int["momentum_x"],
+            results_int["momentum_y"],
+            results_int["continuity"],
+        )
+        pred_u, pred_v, pred_p = results_int["u"], results_int["v"], results_int["p"]
+
+        u_in, v_in = self.parabolic_inflow(self.coords_inflow[:, 1:2])
+
+        # Compute losses
+        # data loss
+        loss_u = torch.mean((self.gnn_u - pred_u) ** 2)
+        loss_v = torch.mean((self.gnn_v - pred_v) ** 2)
+        loss_p = torch.mean((self.gnn_p - pred_p) ** 2)
+
+        # inflow boundary condition loss
+        loss_u_in = torch.mean((u_in - pred_u_in) ** 2)
+        loss_v_in = torch.mean((v_in - pred_v_in) ** 2)
+
+        # noslip boundary condition loss
+        loss_u_noslip = torch.mean(pred_u_noslip**2)
+        loss_v_noslip = torch.mean(pred_v_noslip**2)
+
+        # pde loss
+        loss_mom_u = torch.mean(pred_mom_u**2)
+        loss_mom_v = torch.mean(pred_mom_v**2)
+        loss_cont = torch.mean(pred_cont**2)
+
+        return (
+            loss_u,
+            loss_v,
+            loss_p,
+            loss_u_in,
+            loss_v_in,
+            loss_u_noslip,
+            loss_v_noslip,
+            loss_mom_u,
+            loss_mom_v,
+            loss_cont,
+        )
+
+    def train(self):
+        (
+            loss_u,
+            loss_v,
+            loss_p,
+            loss_u_in,
+            loss_v_in,
+            loss_u_noslip,
+            loss_v_noslip,
+            loss_mom_u,
+            loss_mom_v,
+            loss_cont,
+        ) = self.loss()
+        loss = (
+            1 * loss_u
+            + 1 * loss_v
+            + 1 * loss_p
+            + 10 * loss_u_in
+            + 10 * loss_v_in
+            + 10 * loss_u_noslip
+            + 10 * loss_v_noslip
+            + 1 * loss_mom_u
+            + 1 * loss_mom_v
+            + 10 * loss_cont
+        )
+        self.optimizer.zero_grad()
+        loss.backward()
+        self.optimizer.step()
+
+        return (
+            loss_u,
+            loss_v,
+            loss_p,
+            loss_u_in,
+            loss_v_in,
+            loss_u_noslip,
+            loss_v_noslip,
+            loss_mom_u,
+            loss_mom_v,
+            loss_cont,
+        )
+
+    def validation(self):
+        self.model.eval()
+        with torch.no_grad():
+            x_int, y_int = self.coords[:, 0:1], self.coords[:, 1:2]
+            results_int = self.graph.forward({"x": x_int, "y": y_int})
+            pred_u, pred_v, pred_p = (
+                results_int["u"],
+                results_int["v"],
+                results_int["p"],
+            )
+            pred_u = pred_u.detach().cpu().numpy()
+            pred_v = pred_v.detach().cpu().numpy()
+            pred_p = pred_p.detach().cpu().numpy()
+            error_u = np.linalg.norm(self.ref_u - pred_u) / np.linalg.norm(self.ref_u)
```

Comment:
Can't these all be done on the GPU using torch.norm and only moving the final error values to the CPU?

Reply:
Certainly possible! Just updated it.